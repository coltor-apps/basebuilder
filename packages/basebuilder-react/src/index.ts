export {
  Builder,
  useActiveEntityId,
  useBuilderStore,
  useBuilderStoreData,
} from "./builder";

export {
  Interpreter,
  useInterpreterStore,
  useInterpreterStoreData,
  useEntitiesValues,
} from "./interpreter";

export { createEntityComponent } from "./entities";
export type {
  EntityComponentProps,
  GenericEntityComponent,
  GenericEntityProps,
} from "./entities";

export { createAttributeComponent } from "./attributes";
export type {
  GenericAttributeComponent,
  GenericAttributeProps,
  AttributeComponentProps,
} from "./attributes";
