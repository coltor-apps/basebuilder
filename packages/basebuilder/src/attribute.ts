import { type Schema, type SchemaEntityWithId } from "./schema";

type AttributeContext = {
  schema: Schema;
  entity: SchemaEntityWithId;
};

export type Attribute<TName extends string = string, TValue = unknown> = {
  name: TName;
  validate: (value: unknown, context: AttributeContext) => TValue;
};

export type AttributesValues<
  TAttributes extends ReadonlyArray<Attribute> = ReadonlyArray<Attribute>,
> = {
  [K in TAttributes[number]["name"]]: Awaited<
    ReturnType<Extract<TAttributes[number], { name: K }>["validate"]>
  >;
};

export function createAttribute<const TName extends string, TValue>(
  options: Attribute<TName, TValue>,
): Attribute<TName, TValue> {
  return options;
}
