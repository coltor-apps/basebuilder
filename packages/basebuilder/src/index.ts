export { createBuilder } from "./builder";
export type { Builder } from "./builder";

export { createEntity } from "./entity";
export type { Entity } from "./entity";

export { createAttribute } from "./attribute";
export type { Attribute } from "./attribute";

export type { SubscriptionEvent } from "./subscription-manager";

export { builderStoreEventsNames, createBuilderStore } from "./builder-store";
export type {
  BuilderStore,
  BuilderStoreData,
  BuilderStoreEvent,
  BuilderStoreEventName,
} from "./builder-store";

export {
  createInterpreterStore,
  interpreterStoreEventsNames,
} from "./interpreter-store";
export type {
  InterpreterStore,
  InterpreterStoreData,
  InterpreterStoreEvent,
  InterpreterStoreEventName,
} from "./interpreter-store";

export {
  SchemaValidationError,
  schemaValidationErrorCodes,
  validateSchema,
  validateSchemaIntegrity,
} from "./schema";
export type {
  EntitiesAttributesErrors,
  EntityAttributesErrors,
  Schema,
  SchemaEntity,
  SchemaEntityWithId,
  SchemaValidationErrorCode,
  SchemaValidationErrorReason,
} from "./schema";

export { validateEntitiesValues } from "./entities-values";
export type {
  EntitiesErrors,
  EntitiesValues,
  EntityValue,
} from "./entities-values";
