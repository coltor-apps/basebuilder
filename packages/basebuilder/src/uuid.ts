import cryptoNative from "node:crypto";

export function generateUuid(): string {
  const generator = cryptoNative.randomUUID ?? (() => crypto.randomUUID());

  if (!generator) {
    throw new Error('The "crypto" module is not supported.');
  }

  return generator();
}

export function validateUuid(id: string): void {
  if (
    typeof id !== "string" ||
    !/^(?:[0-9a-f]{8}-[0-9a-f]{4}-[1-5][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}|00000000-0000-0000-0000-000000000000)$/i.test(
      id,
    )
  ) {
    throw new Error(`The entity id '${id}' is invalid.`);
  }
}
