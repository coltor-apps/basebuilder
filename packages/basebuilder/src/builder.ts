import { type Attribute, type AttributesValues } from "./attribute";
import { type Entity } from "./entity";
import { type Schema, type SchemaEntityWithId } from "./schema";
import { generateUuid, validateUuid } from "./uuid";

type ChildrenAllowed<TEntities extends ReadonlyArray<Entity>> = {
  [K in TEntities[number]["name"]]?:
    | ReadonlyArray<TEntities[number]["name"]>
    | true;
};

export type EntitiesExtensions<
  TEntities extends ReadonlyArray<Entity> = ReadonlyArray<Entity>,
> = {
  [K in TEntities[number]["name"]]?: {
    attributes?: {
      [K2 in Extract<
        TEntities[number],
        { name: K }
      >["attributes"][number]["name"]]?: {
        validate?: (
          value: AttributesValues<
            Extract<TEntities[number], { name: K }>["attributes"]
          >[K2],
          context: {
            schema: Schema<Builder<TEntities>>;
            entity: SchemaEntityWithId<
              Builder<[Extract<TEntities[number], { name: K }>]>
            >;
          },
        ) =>
          | AttributesValues<
              Extract<TEntities[number], { name: K }>["attributes"]
            >[K2]
          | Promise<
              AttributesValues<
                Extract<TEntities[number], { name: K }>["attributes"]
              >[K2]
            >;
      };
    };
  };
};

export type Builder<
  TEntities extends ReadonlyArray<Entity> = ReadonlyArray<Entity>,
  TChildrenAllowed extends ChildrenAllowed<TEntities> = Record<
    string,
    true | ReadonlyArray<string> | undefined
  >,
  TParentRequired extends ReadonlyArray<
    TEntities[number]["name"]
  > = ReadonlyArray<string>,
  TEntitiesExtensions extends EntitiesExtensions<TEntities> = EntitiesExtensions<
    []
  >,
> = {
  entities: TEntities;
  childrenAllowed: TChildrenAllowed;
  parentRequired: TParentRequired;
  generateEntityId(): string;
  validateEntityId(id: string): void;
  validateSchema(
    schema: Schema<
      Builder<TEntities, TChildrenAllowed, TParentRequired, TEntitiesExtensions>
    >,
  ):
    | Promise<
        Schema<
          Builder<
            TEntities,
            TChildrenAllowed,
            TParentRequired,
            TEntitiesExtensions
          >
        >
      >
    | Schema<
        Builder<
          TEntities,
          TChildrenAllowed,
          TParentRequired,
          TEntitiesExtensions
        >
      >;
  entitiesExtensions: TEntitiesExtensions;
};

type OptionalBuilderArgs =
  | "childrenAllowed"
  | "parentRequired"
  | "validateSchema"
  | "generateEntityId"
  | "entitiesExtensions"
  | "validateEntityId";

export function createBuilder<
  const TEntities extends ReadonlyArray<Entity>,
  const TChildrenAllowed extends ChildrenAllowed<TEntities> = Record<
    string,
    never
  >,
  const TParentRequired extends ReadonlyArray<TEntities[number]["name"]> = [],
  const TEntitiesExtensions extends EntitiesExtensions<TEntities> = EntitiesExtensions<TEntities>,
>(
  options: Omit<
    Builder<TEntities, TChildrenAllowed, TParentRequired, TEntitiesExtensions>,
    OptionalBuilderArgs
  > &
    Partial<
      Pick<
        Builder<
          TEntities,
          TChildrenAllowed,
          TParentRequired,
          TEntitiesExtensions
        >,
        OptionalBuilderArgs
      >
    >,
): Builder<TEntities, TChildrenAllowed, TParentRequired, TEntitiesExtensions> {
  function fallbackValidateSchema(
    schema: Schema<
      Builder<TEntities, TChildrenAllowed, TParentRequired, TEntitiesExtensions>
    >,
  ): Schema<
    Builder<TEntities, TChildrenAllowed, TParentRequired, TEntitiesExtensions>
  > {
    return schema;
  }

  return {
    ...options,
    validateSchema: options.validateSchema ?? fallbackValidateSchema,
    generateEntityId: options.generateEntityId ?? generateUuid,
    validateEntityId: options.validateEntityId ?? validateUuid,
    entitiesExtensions:
      options.entitiesExtensions ?? ({} as TEntitiesExtensions),
    childrenAllowed: options.childrenAllowed ?? ({} as TChildrenAllowed),
    parentRequired:
      options.parentRequired ??
      ([] as ReadonlyArray<unknown> as TParentRequired),
  };
}

export function getEntityDefinition(
  entityType: string,
  builder: Builder,
): Builder["entities"][number] | undefined {
  return builder.entities.find(
    (builderEntity) => builderEntity.name === entityType,
  );
}

export function ensureEntityIsRegistered(
  entityType: string,
  builder: Builder,
): Builder["entities"][number] {
  const entityDefinition = getEntityDefinition(entityType, builder);

  if (!entityDefinition) {
    throw new Error(`Unkown entity type "${entityType}".`);
  }

  return entityDefinition;
}

export function ensureEntityAttributeIsRegistered(
  entityType: string,
  attributeName: string,
  builder: Builder,
): Attribute {
  const entityDefinition = ensureEntityIsRegistered(entityType, builder);

  const attribute = entityDefinition.attributes.find(
    (attribute) => attribute.name === attributeName,
  );

  if (!attribute) {
    throw new Error(`Unkown entity attribute "${attributeName}".`);
  }

  return attribute;
}

export function ensureEntityAttributesAreRegistered(
  entityType: string,
  attributeNames: Array<string>,
  builder: Builder,
): Array<Attribute> {
  const attributes = attributeNames.map((attributeName) =>
    ensureEntityAttributeIsRegistered(entityType, attributeName, builder),
  );

  return attributes;
}

export function isEntityChildAllowed(
  entityType: string,
  childEntityType: string,
  builder: Builder,
): boolean {
  const allowedChildren = builder.childrenAllowed[entityType];

  if (!allowedChildren) {
    return false;
  }

  return allowedChildren === true || allowedChildren.includes(childEntityType);
}

export function isEntityParentRequired(
  entityType: string,
  builder: Builder,
): boolean {
  return builder.parentRequired.includes(entityType);
}

export function ensureEntityChildAllowed(
  entityType: string,
  childEntityType: string,
  builder: Builder,
): void {
  if (!isEntityChildAllowed(entityType, childEntityType, builder)) {
    throw new Error("Child is not allowed.");
  }
}

export function ensureEntityCanLackParent(
  entityType: string,
  builder: Builder,
): void {
  if (isEntityParentRequired(entityType, builder)) {
    throw new Error("A parent is required.");
  }
}
