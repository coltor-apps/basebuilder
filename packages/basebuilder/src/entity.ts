import { type Attribute, type AttributesValues } from "./attribute";
import { type Builder } from "./builder";
import { type Schema, type SchemaEntityWithId } from "./schema";

export type EntityContext<TAttributes extends ReadonlyArray<Attribute>> = {
  entity: {
    id: string;
    attributes: AttributesValues<TAttributes>;
    children?: Array<string>;
    parentId?: string;
  };
  entitiesValues: Record<string, unknown>;
};

export type AttributesExtensions<TEntity extends Entity = Entity> = {
  [K in TEntity["attributes"][number]["name"]]?: {
    validate?: (
      value: unknown,
      context: {
        schema: Schema;
        entity: SchemaEntityWithId<Builder<[TEntity]>>;
        validate: (
          value: unknown,
        ) => ReturnType<
          Extract<TEntity["attributes"][number], { name: K }>["validate"]
        >;
      },
    ) =>
      | AttributesValues<TEntity["attributes"]>[K]
      | Promise<AttributesValues<TEntity["attributes"]>[K]>;
  };
};

export type Entity<
  TName extends string = string,
  TAttributes extends ReadonlyArray<Attribute> = ReadonlyArray<Attribute>,
  TValue = unknown,
  TParentRequired extends boolean = boolean,
  TChildrenAllowed extends boolean = boolean,
> = {
  name: TName;
  attributes: TAttributes;
  valueAllowed: boolean;
  childrenAllowed: TChildrenAllowed;
  parentRequired: TParentRequired;
  attributesExtensions: AttributesExtensions<Entity<TName, TAttributes>>;
  validate: (value: unknown, context: EntityContext<TAttributes>) => TValue;
  defaultValue: (
    context: EntityContext<TAttributes>,
  ) => Awaited<TValue> | undefined;
  shouldBeProcessed: (context: EntityContext<TAttributes>) => boolean;
};

type OptionalEntityArgs =
  | "attributes"
  | "validate"
  | "defaultValue"
  | "shouldBeProcessed"
  | "childrenAllowed"
  | "attributesExtensions"
  | "parentRequired";

export function createEntity<
  const TName extends string,
  const TAttributes extends ReadonlyArray<Attribute>,
  TValue,
  const TChildrenAllowed extends boolean = false,
  const TParentRequired extends boolean = false,
>(
  options: Omit<
    Entity<TName, TAttributes, TValue, TParentRequired, TChildrenAllowed>,
    OptionalEntityArgs | "valueAllowed"
  > &
    Partial<
      Pick<
        Entity<TName, TAttributes, TValue, TParentRequired, TChildrenAllowed>,
        OptionalEntityArgs
      >
    >,
): Entity<TName, TAttributes, TValue, TParentRequired, TChildrenAllowed> {
  const fallbackAttributes = [] as ReadonlyArray<unknown> as TAttributes;

  function fallbackValidator(value: unknown): TValue {
    if (typeof value !== "undefined") {
      throw new Error(
        `Values for entities of type '${options.name}' are not allowed.`,
      );
    }

    return undefined as TValue;
  }

  function fallbackDefaultValue(): undefined {
    return undefined;
  }

  function fallbackShouldBeProcessed(): boolean {
    return true;
  }

  return {
    ...options,
    childrenAllowed: options.childrenAllowed ?? (false as TChildrenAllowed),
    parentRequired: options.parentRequired ?? (false as TParentRequired),
    attributes: options.attributes ?? fallbackAttributes,
    valueAllowed: typeof options.validate === "function",
    attributesExtensions: options.attributesExtensions ?? {},
    validate: options.validate ?? fallbackValidator,
    defaultValue: options.defaultValue ?? fallbackDefaultValue,
    shouldBeProcessed: options.shouldBeProcessed ?? fallbackShouldBeProcessed,
  };
}
