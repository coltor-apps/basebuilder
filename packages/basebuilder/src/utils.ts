import { type Entity } from "./entity";

export function getEntitiesNamesExcept<
  const TEntities extends ReadonlyArray<Entity>,
  const TExcludeName extends ReadonlyArray<TEntities[number]["name"]>,
>(
  entities: TEntities,
  excludeName: TExcludeName,
): ReadonlyArray<Exclude<TEntities[number]["name"], TExcludeName[number]>> {
  const filteredEntities = entities.filter(function (entity): entity is {
    name: Exclude<TEntities[number]["name"], TExcludeName[number]>;
    attributes: [];
    valueAllowed: boolean;
    childrenAllowed: boolean;
    validate: () => unknown;
    defaultValue: () => unknown;
    shouldBeProcessed: () => boolean;
  } {
    return !excludeName.includes(entity.name);
  });

  if (filteredEntities.length === entities.length) {
    throw new Error(`No such entity "${excludeName}".`);
  }

  return filteredEntities.map((entity) => entity.name);
}

export function insertIntoSetAtIndex<T>(
  set: Set<T>,
  value: T,
  index?: number,
): Set<T> {
  const result = Array.from(set);

  result.splice(index ?? set.size, 0, value);

  return new Set(result);
}

type UndefinedKeys<T> = {
  [K in keyof T]: undefined extends T[K] ? K : never;
}[keyof T];

type NonUndefinedKeys<T> = {
  [K in keyof T]: undefined extends T[K] ? never : K;
}[keyof T];

export type OptionalPropsIfUndefined<T> = {
  [K in UndefinedKeys<T>]?: T[K];
} & {
  [K in NonUndefinedKeys<T>]: T[K];
};

export type KeyofUnion<T> = T extends unknown ? keyof T : never;
