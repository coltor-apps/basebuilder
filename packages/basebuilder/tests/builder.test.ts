import { afterEach, describe, expect, it, vi } from "vitest";

import { createBuilder, createEntity } from "../src";
import * as uuidExports from "../src/uuid";

describe("builder", () => {
  afterEach(() => {
    vi.restoreAllMocks();
  });

  it("can be created with minimal options", () => {
    const generateUuidMock = vi.spyOn(uuidExports, "generateUuid");

    const validateUuidMock = vi.spyOn(uuidExports, "validateUuid");

    const builder = createBuilder({
      entities: [],
    });

    expect(builder).toMatchSnapshot();

    builder.entityId.generate();

    builder.entityId.validate("6e0035c3-0d4c-445f-a42b-2d971225447c");

    expect(generateUuidMock).toHaveBeenCalled();

    expect(validateUuidMock).toHaveBeenCalled();
  });

  it("can be created with entity ID generator and validator", () => {
    const builder = createBuilder({
      entities: [],
      entityId: {
        generate() {
          return "1";
        },
        validate(id) {
          if (id !== "1") {
            throw new Error("Not equal to 1");
          }
        },
      },
    });

    expect(builder).toMatchSnapshot();

    expect(builder.entityId.generate()).toMatchSnapshot();

    expect(builder.entityId.validate("1")).toMatchSnapshot();

    expect(() => builder.entityId.validate("2")).toThrowErrorMatchingSnapshot();
  });

  it("can be created with entities", () => {
    const builder = createBuilder({
      entities: [createEntity({ name: "test" })],
    });

    expect(builder).toMatchSnapshot();
  });

  it("can be created with children and parent rules", () => {
    const builder = createBuilder({
      entities: [createEntity({ name: "test" })],
      childrenAllowed: {
        test: true,
      },
      parentRequired: ["test"],
    });

    expect(builder).toMatchSnapshot();
  });
});
