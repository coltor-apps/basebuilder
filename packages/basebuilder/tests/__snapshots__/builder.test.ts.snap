// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`builder > can be created with children and parent rules 1`] = `
{
  "childrenAllowed": {
    "test": true,
  },
  "entities": [
    {
      "attributes": [],
      "defaultValue": [Function],
      "isValueAllowed": false,
      "name": "test",
      "shouldBeProcessed": [Function],
      "validate": [Function],
    },
  ],
  "entitiesExtensions": {},
  "generateEntityId": [Function],
  "parentRequired": [
    "test",
  ],
  "validateEntityId": [Function],
  "validateSchema": [Function],
}
`;

exports[`builder > can be created with entities 1`] = `
{
  "childrenAllowed": {},
  "entities": [
    {
      "attributes": [],
      "defaultValue": [Function],
      "isValueAllowed": false,
      "name": "test",
      "shouldBeProcessed": [Function],
      "validate": [Function],
    },
  ],
  "entitiesExtensions": {},
  "generateEntityId": [Function],
  "parentRequired": [],
  "validateEntityId": [Function],
  "validateSchema": [Function],
}
`;

exports[`builder > can be created with entity ID generator and validator 1`] = `
{
  "childrenAllowed": {},
  "entities": [],
  "entitiesExtensions": {},
  "generateEntityId": [Function],
  "parentRequired": [],
  "validateEntityId": [Function],
  "validateSchema": [Function],
}
`;

exports[`builder > can be created with entity ID generator and validator 2`] = `"1"`;

exports[`builder > can be created with entity ID generator and validator 3`] = `undefined`;

exports[`builder > can be created with entity ID generator and validator 4`] = `"Not equal to 1"`;

exports[`builder > can be created with minimal options 1`] = `
{
  "childrenAllowed": {},
  "entities": [],
  "entitiesExtensions": {},
  "generateEntityId": [MockFunction generateUuid],
  "parentRequired": [],
  "validateEntityId": [MockFunction validateUuid],
  "validateSchema": [Function],
}
`;
